### DISCUSSION TITLE
Intervention Image as image library

### DISCUSSION THREAD

**Discussion by matteotrubini** [wintercms/winter#141] [Ideas]
EDIT: Plugin available here: https://packagist.org/packages/mercator/wn-media-plugin

We could rely on the "battle tested" [Intervention Image](http://image.intervention.io/) library to provide image handling and manipulation.

In this way we could switch from `gd` to `imagick` using an unified interface.

In the meanwhile:

1. libgd merged support avif -- libgd/libgd#671
2. waiting for php -- php/php-src#5127
3. Intervention Image AVIF Support -- Intervention/image#1056

**Comment by LukeTowers** [comment]
I'm definitely not opposed to this, any thoughts from @bennothommo?

**Reply by bennothommo** [reply]
@LukeTowers @matteotrubini I'm on board with this. My company used Intervention Image for its own image resizer plugin (before the new Resizer functionality was done). It's a very simple API.

**Comment by helmutkaufmann** [comment]
@mjauvin and I are fiddling around with a "gimme the best image format", which returns the best compressed image type based on the browser's capabilities. Developers should not have to think of image types. At the moment, webp is the only "next generation" format available. AVIF as well as HEIF would actually be a welcome extensions.

**Reply by bennothommo** [reply]
@helmutkaufmann That's a great idea for a plugin. The Resizer allows for the extension of the image config, so it should be very easy to implement.

**Reply by helmutkaufmann** [reply]
Can you give me a hint where to start, please, @bennothommo ?

**Reply by bennothommo** [reply]
@helmutkaufmann Sure,

The two events you would want to listen to, to modify the resize functionality are:

- [system.resizer.getDefaultOptions](https://wintercms.com/docs/events/event/system.resizer.getDefaultOptions)
- [system.resizer.processResize](https://wintercms.com/docs/events/event/system.resizer.processResize)

This would allow you to set the default image format to `webp` as part of the default options, and then handle the `webp` resizing in the `processResize` event if they do not override the format with another specific format.

**Reply by helmutkaufmann** [reply]
Ha, thanks, ... That's great, I will look at it later today. Regards to Australia (right you are down under?)

**Reply by bennothommo** [reply]
Yep, we're upside down here :P

**Reply by helmutkaufmann** [reply]
I feel the same at the moment...

**Reply by helmutkaufmann** [reply]
Thank you ... I am again amazed who much one can do with the events. system.resizer.getDefaultOptions works nicely.

**Reply by helmutkaufmann** [reply]
OK. I have - quick and dirty - replaced the Resize with the Intervention library - and it gives the -subjective - impression of being resource-hungry.

**Reply by helmutkaufmann** [reply]
So, Intervention included, you can even daisy-chain commands, e.g. iresize(150, 150, { mode: 'landscape', inter: 'blur(10)->resize(10,10)' }) 

The only challenge I have... the Intervention library is resource-intensive. I would best have to queue all images for creation in the background.

**Reply by helmutkaufmann** [reply]
Update: While the Intervention library provides a nice API, I don't think this is a viable solution as the processing times are beyond what is acceptable, in particular for larger images, which are processed fine with in built-in resizer. I have done some reading and there are quite a lot of people mentioning poor performance. Any other suggestions?

**Reply by helmutkaufmann** [reply]
@bennothommo, I think we can close this discussion, a respective plugin has been built and is available as composer package, see https://packagist.org/packages/mercator/wn-media-plugin.
